keps/sig-storage/20190122-csi-inline-volumes.md

```diff
@@ -276,142 +163,30 @@ To control which driver is allowed to be used within a pod spec, this design wil
 * The Kubelet will create necessary mount point paths
-* Kubelet will auto-generate a volumeHandle based on `podUID` and `pod.spec.volume[x].name` (see above for detail)
-* Ephemeral drivers will receive mount-like calls (NodePublish) with generated paths and volumeHandle
-* Ephemeral drivers are responsible for handling volume operations (create, mount, unmount, delete) during this stage
+* Kubelet will auto-generate a volumeHandle based on `podUID` and `pod.spec.volume[x].name` (see above for detail).
+* CSI driver will receive mount-like calls (NodePublish) with generated paths and generated volumeHandle.
+
+Since ephemeral volume requests will participate in only the mount/unmount volume operation phase, CSI drivers are responsible for implementing all necessary operations during that phase (i.e. create, mount, unmount, delete, etc).  For instance, a driver would be responsible for provisioning any new volume resource during the `NodePublishStage` and for tearing down these resources during the `NodeUnpublish` call.
 
 
 ## Test plans
 
 ### All unit tests
 * Volume operation that use CSIVolumeSource can only work with proper feature gate enabled
 
-### Kubelet persistent inline volumes unit tests
-* Ensure required fields are provided
-* Attachment/detachment should be triggered with CSIVolumeSource
-* Mount/Unmount should be triggered with CSIVolumeSource
-* `CSIDriver.Lifecycle` is set to `persistent` or is empty
-* Validate non empty `volumeHandle`
-* Ensure volumeHandle conforms to resource naming format
-* CSIVolumeSource info persists in CSI json file during attach/detach
-* CSIVolumeSource info persists in CSI json file during mount/unmount
-
-### Kubelet ephemeral inline volumes unit tests
+### Ephemeral inline volumes unit tests
 * Ensure required fields are provided
 * Mount/Unmount should be triggered with CSIVolumeSource
-* `CSIDriver.Lifecycle` is set to `Ephemeral`
-* Ignore `volumeHandle` if not empty
 * Expected generated volumeHandle is created properly
 * Ensure volumeHandle conforms to resource naming format
 * CSIVolumeSource info persists in CSI json file during mount/unmount
 
-### External attacher inline volume unit tess
-* Attachment receives CSIVolumeSource
-* VolumeSource contains a volumeHandle
-* Attach/detach with information from CSIVolumeSource
-
 ### E2E tests
-* Pod spec with a persistent inline volume request can be attached, detached, mount, unmount
 * Pod spec with an ephemeral inline volume request can be mounted/unmounted
-* Two pods accessing the same inline persistent volumes
-* CSI attacher code invokes driver operation during attach for persistent volumes
-* CSI detacher code invokes driver operation during detach for persistent volumes
+* Two pods accessing the same ephemeral inline volumes
```

Comment:
not sure what "the same ephemeral inline volumes" means... the volumes are per pod

Reply:
Say you create a deployment with:
```yaml
volumes:
- name: foo
  csi:
    driver: myephemeraldriver
```

Two pods could land on the same node. If pod A writes to the foo volume, and pod B reads from it, it really should be a different volume between the two, even though they are defined the same in the pod. We should check to ensure that is  the case.

Reply:
ah, verifying an identical inline definition results in distinct volume instances. sounds good.

Reply:
a single pod with two inline ephemeral volumes with identical definitions (other than name) might be easier to test (guaranteed to be on the same node) and test uniqueness down to the subpod level.

Reply:
I think both test cases are important to have.  One validates that the volumehandle is unique per volume name. And the other validates that the volumehandle is unique per pod.

You can use pod affinity to make sure two pods are on the same node.

Reply:
Yeah, both tests are important. While writing the KEP we found problems in the KEP for each case uniquely. The KEP has it fixed now, but should have tests. :)

Reply:
Will update language to indicate what's being tested
1) verifying an identical inline definition results in distinct volume instances
2) a single pod with two inline ephemeral volumes with identical definitions (other than name)